File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
    void generateSubsets(int idx, vector<int>& nums,
                         vector<int>& curr,
                         vector<vector<int>>& ans) {
        // Write your recursive/backtracking code here
        if(idx == nums.size()) {
            ans.push_back(curr);
            return;
        }
        curr.push_back(nums[idx]);
        generateSubsets(idx+1, nums, curr, ans);
        curr.pop_back();
        int next = idx+1;
        while(next < nums.size() && nums[next] == nums[idx]) next++;
        generateSubsets(next, nums, curr, ans);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> arr;
        generateSubsets(0, nums, arr, ans);
        return ans;
    }
};

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    void ps(int[] arr, int i, List<Integer> ans, List<List<Integer>> allSets) {
        if(i == arr.length) {
            allSets.add(new ArrayList<>(ans));
            return;
        }
        ans.add(arr[i]);
        ps(arr, i+1, ans, allSets);
        ans.remove(ans.size()-1);
        int idx = i+1;
        while(idx<arr.length && arr[idx] == arr[idx-1]) idx++;
        ps(arr, idx, ans, allSets);
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
         List<List<Integer>> allSets = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        ps(nums, 0, ans, allSets);
        return allSets;
    }
}